Move RestoreBuildTools to the build traversal from initial targets.

chcosta opened the pull request:
This change moves the restore of build tools from an InitialTarget into the build traversal.  While this somewhat degrades the experience of building a project and solution file in Visual Studio directly from a clean enlistment, it is also a bit less "weird" than the current state which does some "magic" in that experience.  This change is an improvement to the experience from build.cmd / msbuild and reduces weird scenarios like restoring build tools before executing the "clean" target.

It would be nice if we also had an initialtarget which did build tools check validation, and then notified the user to restore build tools if necessary.  Due to internal constraints, we're not at a place where this is feasible.  I believe, though, that this is transient and we'd be able to improve this experience once internal dependencies have been removed.

eatdrinksleepcode commented:
Can you clarify what "weird" and "magic" scenarios this change improves?

> On Apr 15, 2015, at 1:44 PM, chcosta <notifications@github.com> wrote:
> 
> This change moves the restore of build tools from an InitialTarget into the build traversal. While this somewhat degrades the experience of building a project and solution file in Visual Studio directly from a clean enlistment, it is also a bit less "weird" than the current state which does some "magic" in that experience. This change is an improvement to the experience from build.cmd / msbuild and reduces weird scenarios like restoring build tools before executing the "clean" target.
> 
> It would be nice if we also had an initialtarget which did build tools check validation, and then notified the user to restore build tools if necessary. Due to internal constraints, we're not at a place where this is feasible. I believe, though, that this is transient and we'd be able to improve this experience once internal dependencies have been removed.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/dotnet/corefx/pull/1416
> 
> Commit Summary
> 
> Move RestoreBuildTools to the build traversal from initial targets.
> File Changes
> 
> M dir.targets (8)
> M dir.traversal.targets (1)
> M src/dir.targets (9)
> Patch Links:
> 
> https://github.com/dotnet/corefx/pull/1416.patch
> https://github.com/dotnet/corefx/pull/1416.diff
> —
> Reply to this email directly or view it on GitHub.
>

mellinoe commented:
> While this somewhat degrades the experience of building a project and solution file in Visual Studio directly from a clean enlistment

What is the experience from a clean enlistment?

StevenLiekens commented:
@mellinoe from what I can see, the 'Build' target is changed to restore the build tools, then exit. But only when the build tools are not already present. So the experience would be that you have to click Build twice when working from a clean enlistment.

weshaggard commented on src/dir.targets:
```diff
@@ -10,6 +10,15 @@
+    <CallTarget Targets="_RestoreBuildTools" />
+  </Target>
+  <!-- Set a default clean target -->
+  <Target Name="Clean">
```
Why do you need to add the clean here as well?

chcosta replied:
A lot of the current quirkiness comes from being in a clean enlistment or is related to cleaning.  You wouldn't normally expect the RestoreBuildTools step to be tied to "Clean", but it currently is.  

Today's behavior: If you do not have the build tools available (in packages), and you attempt to "Clean" a project (msbuild /t:Clean), it will restore build tools, and then fail with an error message ("error : The build tools package was just restored and so we cannot continue the build of an individual project because targets from the build tools package were not able to be imported. Please retry
 the build the individual project again.").  The second attempt to clean will first run the "RestoreBuildTools" step (which it will skip), and then it will succeed in Cleaning.

The new behavior moves RestoreBuildTools from an InitialTarget, so it no longer runs as part of every msbuild command.  If you type "msbuild /t:Clean", it will just attempt a clean as we would expect.  "msbuild /t:Build" still exhibits the old behavior of requiring you to build twice.

@weshaggard, the reason I'm overriding the "Clean" target is so that you can run "msbuild /t:Clean" from an individual library and it will attempt to clean.  Without overriding the target, you get an annoying "error MSB4057: The target "Clean" does not exist in the project." message.

Of course, you could make the rational statement that you shouldn't be trying to "Clean" a clean build, but I think this scenario does occur where you have a clean repo and then try to clean out of habit.  It's a bit annoying to see an error message in that scenario.  Additionally, if you have previously built binaries, but packages are missing for some reason, you don't want to see an error when trying to clean.

The general guidance for msbuild is that "InitialTargets" are typically used for error checking. I think that making RestoreBuildTools an InitialTarget causes some unexpected behavior and we can have equivalent / sane functionality by making it a precursor to "Build".

weshaggard replied:
Thanks for the explanation @chcosta.

I think everything makes sense to me except for the default behavior of Clean which is to delete BinDir. In my opinion that big hammer only makes sense for the root level build and so I wouldn't do that for leave node projects. So perhaps and empty Clean target is good enough here.

JeremyKuhne replied:
I agree with not wiping the directory.

chcosta replied:
Good feedback.  I originally wanted to clean just specific libraries for the clean command, but found that you didn't always have sufficient behavior, so I resorted to the behavior which I previously noticed for the Clean command of just wiping the bin directory.  I think just having an empty Clean target is better, I may add an informative message in this scenario as well saying that nothing is actually happening.

weshaggard commented on src/dir.targets:
```diff
@@ -10,6 +10,15 @@
 
+  <!-- These gets overwritten when we restore the build tools and import the below targets -->
+  <Target Name="Build" Condition="!Exists('$(ToolsDir)Build.Common.targets')" >
+    <CallTarget Targets="_RestoreBuildTools" />
```
I believe I understand what is happening here but I think we need a better comment here. Something that is mentioning that this Build will only ever run if build tools haven't been restored. It is probably even worth a message here (or perhaps the message already comes out of _RestoreBuildTools in this case), stating that build tools need to be restored.

chcosta replied:
Good point.  Including an additional comment and message here could be useful.

weshaggard commented:
cc @JeremyKuhne

JeremyKuhne commented on dir.traversal.targets:
```diff
@@ -77,6 +77,7 @@
 
   <PropertyGroup>
     <TraversalBuildDependsOn>
+      _RestoreBuildTools;
```
Please also add this to RestoreAllPackages.

chcosta commented:
@dotnet-bot test this please

weshaggard commented on src/dir.targets:
```diff
@@ -10,6 +10,16 @@
+  <!-- This Build target is only run if build tools have not been restored.  Once build tools have been restored, these gets overridden 
+       by the build tools import for the Build target. -->
+  <Target Name="Build" Condition="!Exists('$(ToolsDir)Build.Common.targets')" >
+    <CallTarget Targets="_RestoreBuildTools" />
```
Weren't you going to add a message here about falling back to this target? Or is that message coming out of _RestoreBuildTools itself?

chcosta replied:
I had planned to add an additional message, but then thought that the message coming from restorebuildtools itself was sufficient.

weshaggard commented:
The change looks good to me.

eatdrinksleepcode commented:
I'm still not clear on what the point of this change is. The original PR description mentions the current state being "weird" and doing "magic"; but I don't know what that refers to. If this is just to keep from restoring build tools for the top level clean, then I don't agree with making this change. The bottom line is that if you are going to work with this repo, you need the build tools. Degrading the per-project clean experience from a new enlistment (a lot of developers habitually run "Clean,Build" which would be broken by this change) just to skip a step that you are going to have to do anyway is not an improvement.

chcosta commented:
@eatdrinksleepcode, apologies for not directly addressing your feedback.  I think you bring up some very good points that deserve a response.

I think that anything which happens that was not explicitly intentional, is somewhat "magic".  One of the wonderful things about software, is that it does stuff for us.  When it does more than what we expect, things we had hoped for; that's great, but it should be intuitive.  I, personally, like to understand why things are occurring and think they should make sense as a natural progression of my intended action.  If things just happen, as a by-product of some quirky behavior or an unexpected design side-affect, I think that's weird and I'm hesitant to take a dependency there.  In my opinion, that's what is happening here.  We don't clean a build and expect build tools to appear.  If we build, and there are no build tools, and then the software tells me that it couldn't find build tools so it restored them for me so I should try building again; now it feels intuitive and less "magical".  That being said, I'd prefer this not devolve into a discussion about my poor or obscure wording.  The use of the words was quoted to explicitly highlight the fuzziness of the issue, but I should have explained my use up front.  

To address the more technical aspects of your feedback, while this change does affect the behavior of clean, I don't believe that it affects it detrimentally.  When I say "top-level", I'm referring to building (or cleaning) from the root of the repo, and not a per-project build or clean.  In this scenario, a "clean" from the top-level will not restore build tools. A "build" from the top-level will restore build tools and then build each project, essentially with the same outcome that you'd see without this change.  You would not see a build failure, it would restore build tools, and then build each project.

I don't believe that there is a per-project "clean" experience degradation with this change.  If you clean from a project, without build tools restored, it will be a no-op as opposed to an error.  

If 'msbuild /t:"Clean,Build"' on a single project is your workflow, then that would have a slightly different flow, but provide the same result you see today. The result would be an error the first time, and success the second time.  

If, instead, you're saying that your workflow is "msbuild /t:Clean" followed by "msbuild /t:Build" on a single project, then I'd argue that you're taking a dependency on unintuitive behavior of the architecture.  I'm not certain what the benefit of taking a dependency on a failure occurring during the Clean step which also happens to restore your build tools is. What I'd prefer is a workflow where you can do something like "msbuild /t:Clean", "msbuild /t:RestoreBuildTools", "msbuild /t:Build" and you don't get errors along the way.  In my opinion, this change is closer to enabling a clear flow.

Of course, what would be great is if Clean just did Clean stuff, and Build restored build tools automatically (when needed) and built your project, never requiring a second build.  Unfortunately, the architecture of MsBuild makes that simple desire difficult to obtain in an intuitive manner, so we have this state where we, on some occasions, have to build twice; but I hope that we can be clear to the user about what is happening and take some of the weird magic out of it. :)

StevenLiekens commented:
There's something weird about using MSBuild targets to restore other MSBuild targets. It doesn't really matter how you refactor this. It will always be weird.

ellismg commented:
@dotnet-bot test this please

eatdrinksleepcode commented:
Thanks @chcosta for taking the time to explain your perspective. You are still using highly subjective language to describe your reasoning, which is understandable since this is a highly subjective issue.

| We don't clean a build and expect build tools to appear.
Why not? My mental model of the relationship between BuildTools and the CoreFx build scripts has been one of base class and derived class. In order to create an instance of a derived class, the base class has to be available, even if you don't intend on using any functionality from the base class. Similarly, it makes perfect sense to me that the BuildTools have to be available before ANY target in the CoreFx build runs, even if that target does not require anything from the BuildTools.

eatdrinksleepcode commented:
Thanks @chcosta for taking the time to explain your perspective. You are still using highly subjective language to describe your reasoning, which is understandable since this is a highly subjective issue.

> We don't clean a build and expect build tools to appear.

Why not? My mental model of the relationship between BuildTools and the CoreFx build scripts has been one of base class and derived class. In order to create an instance of a derived class, the base class has to be available, even if you don't intend on using any functionality from the base class. Similarly, it makes perfect sense to me that the BuildTools have to be available before ANY target in the CoreFx build runs, even if that target does not specifically require anything from the BuildTools. To me this makes the relationship between the CoreFx build and the BuildTools easier to reason about; if I want to use something from BuildTools in CoreFx, I don't have to worry about whether or not they have been restored or whether I need to add a CallTarget somewhere to get them restored.

> If you clean from a project, without build tools restored, it will be a no-op as opposed to an error.

This to me is far more unintuitive than the previous behavior. It makes a lot more sense to me for the Clean task to say "I have made a little progress but I cannot continue, please try again" than for it to say "you need to run a completely unrelated target first before you can call me".

> If 'msbuild /t:"Clean,Build"' on a single project is your workflow, then that would have a slightly different flow, but provide the same result you see today. The result would be an error the first time, and success the second time.

But the messages you get during that new flow are confusing. First you see that you need to restore the build tools, then you see the build tools restoring, and then an error that build tools have been restored so you cannot continue:

> Clean:
>  'Clean' target is empty. Restore build tools to enable 'Clean' target.
> _RestoreBuildTools:
>  Restoring build tools...
> ...
> c:\Users\danelson\Projects\Personal\dotnet\corefx\dir.targets(54,5): error : The build tools package was just restored and so we cannot continue the build of an individual project because targets from the build tools package were not able to be imported. Please retry the build the individual project again. [c:\Users\danelson\Projects\Personal\dotnet\corefx\src\System.ComponentModel\src\System.ComponentModel.csproj]

The current experience is much cleaner and, in my opinion, more intuitive.

I realize this is a highly subjective topic, and if the general consensus is that this is an improvement, I will abide by that consensus. But in my opinion  this change represents a degradation in both the usability and maintainability of the build, and it should not be merged.

chcosta commented:
Thanks @eatdrinksleepcode, interesting perspective.  I agree that this is largely dependent on viewpoint and I'm happy that we are able to have this discussion. 

All things being equal, I'd rather not take a subjective change if it is not an improvement or a step towards an incremental improvement.  I'll discuss this change further with the framework architects.  If this change does not enable other scenarios or provide a foundation for the direction we see the build going, then I'd opt not to take this change.

weshaggard commented:
Thanks as always for the great feedback @eatdrinksleepcode but I think the automatic restore build tools approach is starting to fall down in too many scenarios (working in VS is one of the hidden broken scenarios right now) and I suspect we are moving into a direction where you will need to do an explicit restore build tools step after you do any syncing or working directory changes (i.e. change branches). I don't think we are fully ready to force into that directly yet but @chcosta's change allows us to move closer in that direction and test out the waters a little.

As for dev workflows from my experience I've seem many more folks use rebuild then an explicit clean/build in which case this change works the same. That said it may be worth explicitly calling _RestoreBuildTools from the clean target as well given that it is a common target to trigger but I do think it we should move away from having this as an InitialTargets because we are at least going to have a target that allows folks to explicitly restore build tools and other similar type targets that we don't want this initial target tripping up on.

ellismg commented:
> I think the automatic restore build tools approach is starting to fall down in too many scenarios (working in VS is one of the hidden broken scenarios right now) and I suspect we are moving into a direction where you will need to do an explicit restore build tools step after you do any syncing or working directory changes (i.e. change branches).

I don't agree with this approach.  I run git clean -dxf and change branches a bunch in my daily workflow and having to remember to run a command each time is painful.  I understand that most of the time, things will probably just work (if I don't clean), but it still seems like we're moving further away from the ideal world, not closer to it.

I understand the rationale for having binaries we need as part of the build restored via NuGet and in the build tools package.  I don't understand the rationale behind having the targets handled this way.  What if the targets were checked in?  A build tools update would mean bumping the version number for the build tools nuget package (to download the actual binaries we need to build) and updating the targets files that are checked in to the tree.

If we did this, would we be able to stop this craziness of figuring out how to hook VS and MsBuild so all the targets came it at the right time?  What would we be giving up?  It seems to me a script which just took build tools and did these updates would not be painful for repo owners.

We would also get the targets files under source control which would help with finding things via git grep.

What do we gain by having the targets restored via nuget?  Is it worth the cost we're paying?

JeremyKuhne commented:
@ellismg To me, the gains we get from sharing the targets in the BuildTools outweighs the negatives we get from poor package restore tooling support (in msbuild/vs). I'm particularly concerned about how the targets that are checked in are already diverging- I would expect that we would eventually end up with a whole bunch of wildly different build environments if we didn't strive to keep most of this coming from a common source.

JeremyKuhne commented:
@eatdrinksleepcode/@chcosta: 

There is no way to reliably do an in-situ restore of imports if you're directly accessing a project from msbuild. Right now some scenarios "kindof" work because we condition the imports on existance. Signing, references, etc. don't pull in for VS background builds which gives busted intellisense and other strange errors. In addition, going with an initial target solution as we currently do or trying to insert in front of build like this attempts creates issues with parallel builds that aren't easily solved.

The only reliable way to do this is inject from a command line build (build.cmd) and potentially from solution package restore (exactly how that works isn't completely clear- items in the solution level packages.config don't seem to restore automatically with some simple tests I've run).

I think the error condition should be the initial target, and shouldn't be this:

```xml
    <Error Condition="'$(ErrorIfBuildToolsRestoredFromIndividualProject)'=='true'"
           Text="The build tools package was just restored and so we cannot continue the build of an individual project because targets from the build tools package were not able to be imported. Please retry the build the individual project again." />
```

It should be something like:

```xml
    <Error Condition=="!Exists('$(ToolsDir)Build.Common.targets')"
           Text="The build tools package does not exist. Build the project via build.cmd or run "build.cmd restorebuildtools" first." />
```

As I infer in the message above the build.cmd should expose the build tool restore and should handle being invoked with a .csproj/.sln and first invoke the restorebuildtools before passing through to msbuild.

I also believe Clean shouldn't wipe out the package directory and should work without the build package. I think it's reasonable to expect our build tools to not introduce new clean behavior.

That's my opinion anyway after playing with various options over the past several weeks. If you have workflows that this busts I'd love to hear them and discuss how we avoid unnecessary grief yet still provide a dependable build.

eatdrinksleepcode commented:
My ideal build system is one which lets me do a fresh checkout and then execute any build target with the expectation that it will complete successfully. This also dovetails with @ellismg's description of what happens when you need to move around a lot within the commit history. The developer efficiency that comes from not having to think about your build system is, in my experience, under-appreciated. And there is no reason given the powerful build tools that exist today that we should feel the need to settle for a sub-standard developer experience.

That being said, MSBuild has known architectural limitations around bootstrapping its own dependencies. If we have exhausted all reasonable attempts to workaround this limitation while still maintaining a positive developer experience for all common development scenarios (e.g. Visual Studio/command line, clean/build/rebuild, etc.), then I would be in favor of removing the bootstrapping entirely and requiring that the build tools be restored as a separate step. This would be annoying for the reasons that @ellismg outlined, but it might lead to a more intuitive developer experience. It is not clear to me whether we have exhausted those attempts, because I am still not clear on exactly what scenarios are not working with the current setup; this thread has a lot of vague references without much detail. It might help if those could be enumerated.

However, if we are required to make that compromise because of the limitations of the tool, then I also think we ought to strongly consider whether we are using the right tool for the job.

To the question raised about what we gain from putting the targets in another repository: I agree with @JeremyKuhne that there is value in constraining repository drift by centralizing and promoting "good" build practices. However, I also think that we would be presumptuous to claim at this point that we understand what "good" build practices really are under these circumstances with these tools (this thread itself is evidence of that). It might make sense to do a little more of the experimentation with the build in individual repos, and only extract patterns into build tools when they have been proven to be successful. This would make updating the build easier while we are still trying different solutions.

> I also believe Clean shouldn't wipe out the package directory...

I agree with this; Clean should clean the output of the build, not the input (although a target specifically for removing the packages directory is not unreasonable).

> ...and should work without the build package.

Other than a philosophical argument, I don't see a compelling reason from a usage perspective why we should impose this limitation on ourselves.

> I think it's reasonable to expect our build tools to not introduce new clean behavior.

I don't understand what you mean by this. What new clean behavior are the build tools introducing? Or are you saying that it's ok not to restore the build tools during clean because it should never matter? Again, I don't think putting the build through contortions to satisfy that arbitrary condition gains us anything. Maybe the build tools will never modify the clean behavior, or maybe we will find a compelling reason why they should. Regardless, why should NOT restoring build tools during clean be a design goal for the build system? What compelling advantage is gained for developers by achieving that goal?

> If you have workflows that this busts I'd love to hear them and discuss how we avoid unnecessary grief yet still provide a dependable build.

Nothing is busted per se, but I think I described in my previous post why the build behavior with this PR is suboptimal, as did @ellismg. I don't think I have the right answers yet, but that's partly because we haven't really discussed the problem space. This PR just showed up out of nowhere causing (in my opinion) a degradation of the command line build process. I am just one voice here (albeit a loud and persistent one :grin:), but I would like to understand the problem we are trying to solve before making a compromise to solve it.

weshaggard commented:
@ellismg if we did check-in our targets files we would simply hit the next level issues which would be the need to conditionally reference our task library and then we would error out when we hit the calls to those tasks. I agree with @JeremyKuhne and @eatdrinksleepcode that we should have as much as we can in the common buildtools to try and ensure consistent builds and share work across our different repos. The bootstrapping problem is however a challenge that we need to figure out the best ways to handle. 

@chcosta and @JeremyKuhne can you guys work together to list out the set of requirements/scenarios for the builds (i.e. building in VS/commandline from the root or leaf project, getting intellisense/designer builds, switching build tools versions while VS is running, etc) and then list the current state of the build system. I think that will get at what @eatdrinksleepcode is asking about and help clearly identify what issues we are solving with these changes. It will also help us point out any known limitations so folks are aware. This list would actually be useful to go somewhere in our buildtools or corefx wiki under build design so folks can reference it to better understand how our infrastructure works.

JeremyKuhne commented:
@eatdrinksleepcode The experience if you open VisualStudio directly on a project or solution in an enlistment that doesn't have a package directory is one of the "weird" situations we're talking about.

VS loads the projects, but they don't have the build tools imports. References that are resolved from the package dependencies in the build tools won't be available. VS does a number of design time builds in the main process on this partially complete project, including an Intellisense build. Currently that kicks a build tools restore due to the initial targets in whatever design time build happens to run first (which doesn't appear to be deterministic). Any messages that are pushed as part of these builds do not show up in the output/errors pane

If you build the solution/project an out of proc build is run, which will complete successfully as everything is in place when that build engine loads the projects. At that point we have no way of telling that the tools were just restored so we can't give any feedback either. Intellisense (at least) will probably be busted and stay busted until you reload the project.

In my mind this is really undesirable as you need to know way too much about how our build targets and VS works to intuit that you're in this state. I'd much rather not implicitly restore with initial targets and give a hard error in a "normal" build so the user is fully aware of the bad state they are in and what they need to do to correct it. So on build (from VS) you would get something "The build tools were just restored and the build cannot continue successfully, please reload the solution/project to get full VS tooling support."

We could keep the initial targets and condition out known cases as we discover them (!DesignTime being the first one), but I worry about how well we can identify these and how complicated the conditions might get.

The other problem we have with initial targets is if builds are done in parallel it is essentially impossible to avoid collision on the restore. One possibility might be to have a boot-strap project that does this that all solutions contain and have first in order, but that still doesn't solve the stale Project instances inside of VS of course. And this wouldn't solve building an individual project with multiple project references unless building reference projects implicitly is turned off.

The complexity we have to add to make things mostly seamless and robust worries me and pushes me towards "manual step that always works", particularly because this step isn't something you need to do often.

On the "clean" bit I'm open to your opinion there- I'll think about it some more. It may just be a knee-jerk reaction for me- pulling down bits during clean "feels" dirty. :)

Moving around the history doesn't bother me too much. If we're not wiping the packages and they're versioned correctly this shouldn't be an issue should it? What am I missing there?

chcosta commented:
I think this discussion is slightly divergent from the PR itself.  The PR, as it stands today, moves the Restore Build Tools step out of the Initial Targets and into the Build target, **Clean** target, and the Restore Packages target.  

I believe that initial targets should be used for error checking / validation, not for actual work such as restoring build tools.  Given that I believe this PR maintains a level of compat with the current behavior, yet provides more control over the execution of the restore build tools target, I would like to merge this PR.  I think that the discussion about our build tools / targets, the proper experience is very important and I'd like to continue that discussion, but perhaps tracked under a different issue.  I do plan to follow up on Wes' proposed work-item to document the current state of things, and I would like to see further collaboration to improve this area of the build tools.

eatdrinksleepcode commented:
@chcosta If I understand correctly (and I'll be honest, I'm not entirely sure that I do, but I'm trying to keep up), this PR does not improve any of the scenarios (particularly related to Visual Studio) that are of concern in our current state. Is that correct? If so I am not sure what benefit there is in merging this. If it does improve one of the troublesome scenarios (please describe), and that is considered to have greater value than the degradation of the workflow that @ellismg described, then I am not opposed to merging this.

> I believe that initial targets should be used for error checking / validation, not for actual work such as restoring build tools.

That seems to be a philosophical argument. I don't see how restoring the BuildTools in a CallTarget instead of an InitialTarget significantly affects either the maintainability of the build or the developer experience. I would argue (as I suppose I already have) that if anything, centralizing the restoration of the BuildTools makes maintaining the build easier. If we decide that restoring the BuildTools from inside the build is simply not viable due to the limitations of MSBuild and Visual Studio, then we should go ahead and take it out. But it is not clear to me what we gain by moving it around inside the build (if that is in fact all we are doing here).

chcosta commented:
@eatdrinksleepcode, @JeremyKuhne  spoke directly to some of the Visual Studio related scenarios affected by having restore build tools in initial targets.  

I don't see how this change specifically relates to the degradation @ellismg described.  I read ellismg's comment more as a philosophical argument about how we should be handling build tools, not as specific feedback that this PR is degrading the workflow.

eatdrinksleepcode commented:
@JeremyKuhne I haven't seen the kinds of failures you describe. But I also tend to run a command line build after a pull as a sanity check before opening VS, so I am probably working around them without meaning to.

> The complexity we have to add to make things mostly seamless and robust worries me and pushes me towards "manual step that always works", particularly because this step isn't something you need to do often.

This is not unreasonable; it's just really sad. I am not forced to restore the dependencies of the build prior to actually running the build when I use Rake, Gradle, or even Maven (not to mention that all of these tools understand multi-module builds and don't require workarounds for them). Coming back to .NET after working in Java at my day job is like a breath of fresh air, _except_ where build and dependency management is concerned; we are still so far behind in these areas.

> Moving around the history doesn't bother me too much. If we're not wiping the packages and they're versioned correctly this shouldn't be an issue should it? What am I missing there?

@ellismg mentioned that he often runs "git clean -dxf" in order to get to a pristine state as he navigates between commits. Today, running "msbuild" from the root immediately after such a clean succeeds; if we switch to requiring the build tools to be restored first, it won't. This would be mitigated if Nuget packages were referenced from a user- or machine-level cache rather than from a local packages directory, a la Gradle/Maven/Ruby; _or_ if our incremental build were reliable enough that @ellismg did not feel the need for a "git clean" :grin:.

eatdrinksleepcode commented:
> @eatdrinksleepcode, @JeremyKuhne spoke directly to some of the Visual Studio related scenarios affected by having restore build tools in initial targets.

But are these scenarios actually improved by this change? The build tools are still being restored during the build, it's just happening inside of a CallTarget instead of an InitialTarget. Does that actually matter? I was under the impression that it was the initial build inside Visual Studio that is breaking and causing the subsequent problems; such a build will break after this change just like it would before.

JeremyKuhne commented:
@eatdrinksleepcode 
>I haven't seen the kinds of failures you describe. But I also tend to run a command line build after a pull as a sanity check before opening VS, so I am probably working around them without meaning to.

I usually do the same thing, and in most places I've worked/do work.

>This is not unreasonable; it's just really sad.

Totally agree. I've struggled with this for years, having written the build system for Expression on MSBuild. One of the most fundamental workflow problems is that VS doesn't use MSBuild for solution build management. Moving the .sln to MSBuild would go a long way towards solving workflow issues. (Imagine being able to put a target in at the beginning of a solution load, exposing your own solution targets such as "Localize..." in the menus, etc.) Making the solution file a build file has been on the backlog for years, I realllllly hope it (or the equivalent) makes it with Dev15 and I'll continue to take every opportunity to plead for it. :)

>Today, running "msbuild" from the root immediately after such a clean succeeds; if we switch to requiring the build tools to be restored first, it won't.

"Build.cmd" does, is that enough?

>This would be mitigated if Nuget packages were referenced from a user- or machine-level cache rather than from a local packages directory, a la Gradle/Maven/Ruby; or if our incremental build were reliable enough that @ellismg did not feel the need for a "git clean" :grin:.

I'm actually looking into caching like you're talking about. I'm investigating one of the internal build systems that has a global cache for NuGet packages to see what I can leverage, even conceptually. I'll look at the experience with the others you mention as well.

Reliable incremental builds are super important and I'll certainly make every effort to fix anything I run across. :) Unfortunately it is really hard to get right, particularly if you leave implicit builds of project references turned on. :/

Thanks for all of the constructive feedback, btw. I hope you're prepared for me to ping you on the requirements/scenarios write-up around this that Wes mentioned. ;)

eatdrinksleepcode commented:
> Imagine being able to put a target in at the beginning of a solution load, exposing your own solution targets such as "Localize..." in the menus, etc.

This would be amazing, and would move us a lot closer to the Intellij + Gradle experience. Do you know if the MSBuild team is actively working on solution support? Is this something the community could participate in? I have been watching the MSBuild repo, but the team seems to still be struggling with how to engage with the community.

> > Today, running "msbuild" from the root immediately after such a clean succeeds; if we switch to requiring the build tools to be restored first, it won't.

> "Build.cmd" does, is that enough?

Perhaps. I generally avoid shell scripts for running the build; it is difficult for them to expose the full functionality offered by the build script, and they have to be rewritten for each platform. But in this case, if the shell script does nothing but restore the build tools and then invoke the build with whatever parameters are passed to it, that might be a reasonable mitigation.

> I'm actually looking into caching like you're talking about.

I have a lot of experience with this kind of dependency management on other platforms, and I would be very interested in trying to achieve something similar for CoreFx (and eventually other projects). If there is anything I can do to help make this happen, let me know.

> Reliable incremental builds are super important...Unfortunately it is really hard to get right, particularly if you leave implicit builds of project references turned on. :/

They are very difficult to get right, but worth the effort on any non-trivial project. Implicit builds of project references are actually the major reason why incremental build is so important. In Maven when I build a particular project, it doesn't build any of the project references first, which often results in build failures for the current project. As a result, people using Maven tend to get in the habit of always building from the root; but because incremental build in Maven is practically non-existent, root builds tend to take a lot of time, even when there is very little to do. This is wasteful, and frustrating to me as a developer.

I want building a project to always succeed unless the developer has actually broken something, which means referenced projects need to be built first. And I want the build to only do the work that it actually has to do, so that even building a long reference chain is as efficient as possible. Gradle excels at this kind of complete but extremely efficient incremental build; I would highly recommend using it as inspiration for any build you write.

> I hope you're prepared for me to ping you on the requirements/scenarios write-up around this that Wes mentioned.

I am passionate about creating a good developer experience, and the build is a major factor in that experience. Try to keep me away :grin:

chcosta commented:
I appreciate all of the feedback.  Given the lack of consensus, I'm going to close this PR.
